### PR TITLE
daemon: add configuration option for the GC collection interval

### PR DESCRIPTION
**Pull request by nirmoy**
Related: #4400

Signed-off-by: Nirmoy Das <ndas@suse.de>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4529)

### CODE REVIEW

File: daemon/main.go [hunk 1]
```diff
@@ -326,6 +326,7 @@ func init() {
 	flags.StringSliceVar(&debugVerboseFlags, argDebugVerbose, []string{}, "List of enabled verbose debug groups")
 	flags.StringVarP(&option.Config.Device,
 		"device", "d", "undefined", "Device facing cluster/external network for direct L3 (non-overlay mode)")
+	flags.IntP("gc-interval", "", 60, "Garbage collection interval in seconds")
```
Comment (aanm) [inline daemon/main.go]: change `gc-interval` to something like `connection-tracking-garbage-collector-interval` (My proposal is too long, but something should be something more explicit than `gc-interval`. The same goes for the flag description, something like `Garbage collection interval for the connection tracking table (in seconds).

Comment (aanm) [inline daemon/main.go]: The flags are in alphabetical order (at least they should be) So this should be between line 347 and 348

File: pkg/endpointmanager/conntrack.go [hunk 2]
```diff
@@ -80,7 +75,7 @@ func RunGC(e *endpoint.Endpoint, isLocal, isIPv6 bool, filter *ctmap.GCFilter) {
 }
 
 // EnableConntrackGC enables the connection tracking garbage collection.
-func EnableConntrackGC(ipv4, ipv6 bool) {
+func EnableConntrackGC(ipv4, ipv6 bool, GcInterval int) {
```
Comment (aanm) [inline pkg/endpointmanager/conntrack.go]: you can change this variable to be lowercase

**Review by aanm — APPROVED** [review]

**Review by aanm — CHANGES_REQUESTED** [review]
Previous review was for request changes

**Comment by aanm** [comment]
test-me-please

File: pkg/endpointmanager/conntrack.go (hunk 2 above)
Comment (nirmoy) [reply pkg/endpointmanager/conntrack.go]: Thanks a lot reviewing, updated my PR

File: pkg/endpointmanager/conntrack.go [hunk 3]
```diff
@@ -80,10 +75,10 @@ func RunGC(e *endpoint.Endpoint, isLocal, isIPv6 bool, filter *ctmap.GCFilter) {
 }
 
 // EnableConntrackGC enables the connection tracking garbage collection.
-func EnableConntrackGC(ipv4, ipv6 bool) {
+func EnableConntrackGC(ipv4, ipv6 bool, gcinterval int) {
 	go func() {
```
Comment (aanm) [inline pkg/endpointmanager/conntrack.go:87]: Sorry, I just realized something, add an `if` to check if `gcinterval` is less than 1. If it is, then assign `1` to `gcinterval`, just to prevent user to set up the GC to run every second. Maybe 5? defer to @joestringer

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/endpointmanager/conntrack.go (hunk 3 above)
Comment (joestringer) [reply pkg/endpointmanager/conntrack.go:87]: Yeah, I think anything less than 5 is pretty unwise.

**Comment by nirmoy** [comment]
@joestringer Added a check to verify if GcInterval < MinGcInterval(5 sec)

File: daemon/main.go [hunk 4]
```diff
@@ -735,7 +736,7 @@ func runDaemon() {
 	}
 
 	log.Info("Starting connection tracking garbage collector")
-	endpointmanager.EnableConntrackGC(!option.Config.IPv4Disabled, true)
+	endpointmanager.EnableConntrackGC(!option.Config.IPv4Disabled, true, viper.GetInt("conntrack-garbage-collector-interval"))
```
Comment (manalibhutiyani) [inline daemon/main.go:739]: Since this is a user input, we should do some more testing too, like a negative value?

File: pkg/endpointmanager/conntrack.go (hunk 2 above)
Comment (aanm) [reply pkg/endpointmanager/conntrack.go]: This remains unchanged :)

**Review by aanm — CHANGES_REQUESTED** [review]

File: daemon/main.go (hunk 4 above)
Comment (nirmoy) [reply daemon/main.go:739]: Thanks for reviewing, updated to use flags.UintP

**Comment by nirmoy** [comment]
@aanm I missed that :/ updated the PR.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by nirmoy** [comment]
updated docs with make -C Documentation cmdref

File: daemon/main.go [hunk 5]
```diff
@@ -317,6 +317,7 @@ func init() {
 		"bpf-root", "", "Path to BPF filesystem")
 	flags.StringVar(&cfgFile,
 		"config", "", `Configuration file (default "$HOME/ciliumd.yaml")`)
+	flags.UintP("conntrack-garbage-collector-interval", "", 60, "Garbage collection interval for the connection tracking table (in seconds)")
```
Comment (joestringer) [inline daemon/main.go]: Any particular reason to use `flags.UintP()` rather than plain old `flags.Uint()`? Seems like the former allows a shorthand, but this code isn't introducing a shorthand variation (and I don't think it really needs to).

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

File: daemon/main.go (hunk 5 above)
Comment (nirmoy) [reply daemon/main.go]: Thanks for such a close look, I fixed it.

**Comment by aanm** [comment]
test-me-please

File: pkg/endpointmanager/conntrack.go [hunk 6]
```diff
@@ -83,9 +83,13 @@ func RunGC(e *endpoint.Endpoint, isIPv6 bool, filter *ctmap.GCFilter) {
...
-		sleepTime := time.Duration(GcInterval) * time.Second
+		if gcinterval < MinGcInterval {
+			gcinterval = MinGcInterval
+			log.Warnf("Setting conntrack garbage collector interval to its minimum value(%d seconds)", gcinterval)
+		}
+		sleepTime := time.Duration(gcinterval) * time.Second
```
Comment (raybejjani) [inline pkg/endpointmanager/conntrack.go:92]: the `time.Duration` cast is unnecessary, I think. We do have it all over but it would be nice to not add more :/

**Review by raybejjani — APPROVED** [review]

File: pkg/endpointmanager/conntrack.go (hunk 6 above)
Comment (nirmoy) [reply pkg/endpointmanager/conntrack.go:92]: I get invalid "operation: gcinterval * time.Second (mismatched types int and time.Duration) " 
when I use gcinterval* time.Second
(go version go1.10.2 linux/amd64)

Comment (raybejjani) [reply pkg/endpointmanager/conntrack.go:92]: hmm. I guess I was wrong. The other instances I have in mind must have been literal numbers, and they were converted appropriately. Ignore me!
